D3D12HelloWorld/Graphics/CommandList.cs

```diff
@@ -85,23 +85,20 @@ public sealed class CommandList : IDisposable {
             GetCommandQueue().ExecuteCommandLists(Close());
         }
 
-        public void Reset() {
-            mCommandAllocator.Reset();
-            mCommandList.Reset(mCommandAllocator, null);
-            mIsCommandListClosed = false;
-            SetDescriptorHeaps(mDevice.ShaderVisibleShaderResourceViewAllocator, mDevice.ShaderVisibleSamplerAllocator);
-        }
+        public void Reset(ID3D12CommandAllocator? commandAllocatorOverride = null) {
+            ID3D12CommandAllocator commandAllocator = commandAllocatorOverride ?? mCommandAllocator;
 
-        [Obsolete("Work to refactor off this in favour of Reset() and SetPipelineState(PipelineState), which only uses our own allocator, and which sets the root signature according to the pipeline state.")]
-        public void Reset(ID3D12CommandAllocator commandAllocator, PipelineState pipelineState) {
-            mCommandList.Reset(commandAllocator, pipelineState.NativePipelineState);
-            mIsCommandListClosed = false;
-        }
+            // Command list allocators can only be reset when the associated
+            // command lists have finished execution on the GPU; apps should use
+            // fences to determine GPU execution progress.
+            commandAllocator.Reset();
 
-        [Obsolete("Work to refactor off this in favour of SetPipelineState(PipelineState), which sets the root signature according to the pipeline state, and which sets the descriptor heaps for the shaders and samplers managed by GraphicsDevice.")]
-        public void SetNecessaryState(ID3D12RootSignature rootSignature, params ID3D12DescriptorHeap[] descriptorHeaps) {
-            mCommandList.SetGraphicsRootSignature(rootSignature);
-            mCommandList.SetDescriptorHeaps(descriptorHeaps);
```

Comment:
This hack has been removed by exposing new `SetRootSignature` and `SetShaderVisibleDescriptorHeaps` methods to give finer grained control...